pupil_src/shared_modules/calibration_routines/data_processing.py

```diff
@@ -0,0 +1,200 @@
+    except IndexError:
+        return idx - 1
+
+
+def match_data(g_pool, pupil_list, ref_list):
+    """
+    Returns binocular and monocular matched pupil datums and ref points. Uses a dispersion criterion to
+    dismiss matches which are too far apart.
+    """
+    if pupil_list and ref_list:
+        pass
+    else:
+        not_enough_data_error_msg = (
+            "Not enough ref point or pupil data available for calibration."
+        )
+        logger.error(not_enough_data_error_msg)
+        return {
+            "subject": "calibration.failed",
+            "reason": not_enough_data_error_msg,
+            "timestamp": g_pool.get_timestamp(),
+            "record": True,
+        }
+
+    pupil0 = [p for p in pupil_list if p["id"] == 0]
+    pupil1 = [p for p in pupil_list if p["id"] == 1]
+
+    matched_binocular_data = closest_matches_binocular(ref_list, pupil_list)
+    matched_pupil0_data = closest_matches_monocular(ref_list, pupil0)
+    matched_pupil1_data = closest_matches_monocular(ref_list, pupil1)
+
+    if len(matched_pupil0_data) > len(matched_pupil1_data):
+        matched_monocular_data = matched_pupil0_data
+    else:
+        matched_monocular_data = matched_pupil1_data
+
+    logger.info(
+        "Collected {} monocular calibration data.".format(len(matched_monocular_data))
+    )
+    logger.info(
+        "Collected {} binocular calibration data.".format(len(matched_binocular_data))
+    )
+    return (
+        matched_binocular_data,
+        matched_monocular_data,
+        matched_pupil0_data,
+        matched_pupil1_data,
+        pupil0,
+        pupil1,
+    )
+
+
+def closest_matches_binocular(ref_pts, pupil_pts, max_dispersion=1 / 15.0):
+    """
+    get pupil positions closest in time to ref points.
+    return list of dict with matching ref, pupil0 and pupil1 data triplets.
+    """
+    pupil0 = [p for p in pupil_pts if p["id"] == 0]
+    pupil1 = [p for p in pupil_pts if p["id"] == 1]
+
+    pupil0_ts = np.array([p["timestamp"] for p in pupil0])
+    pupil1_ts = np.array([p["timestamp"] for p in pupil1])
+
+    matched = []
+
+    if pupil0 and pupil1:
+        for r in ref_pts:
+
+            closest_p0_idx = find_nearest_idx(pupil0_ts, r["timestamp"])
+            closest_p0 = pupil0[closest_p0_idx]
+            closest_p1_idx = find_nearest_idx(pupil1_ts, r["timestamp"])
+            closest_p1 = pupil1[closest_p1_idx]
+
+            dispersion = max(
+                closest_p0["timestamp"], closest_p1["timestamp"], r["timestamp"]
+            ) - min(closest_p0["timestamp"], closest_p1["timestamp"], r["timestamp"])
+            if dispersion < max_dispersion:
+                matched.append({"ref": r, "pupil": closest_p0, "pupil1": closest_p1})
+            else:
+                logger.debug(
+                    "Binocular match rejected due to time dispersion criterion"
+                )
+
+    return matched
+
+
+def closest_matches_monocular(ref_pts, pupil_pts, max_dispersion=1 / 15.0):
```

Comment:
The code of this function is very redundant with the binocular version. They can probably be merged or redundant code can be extracted.